File: reference/vector_samples_19_4_10/SIL/SilAdapterBasicPython/SUT/Application_Python/logic.py
from iolayer import State, get_sensor_value1, get_sensor_value2, get_sensor_value3, set_heater_state

HEATING_THRESHOLD = 18.0
COOLING_THRESHOLD = 23.0

write_to_console_counter = 0

def get_heater_state(temperature: float) -> State:
    if temperature < HEATING_THRESHOLD:
        return State.Heating
    if temperature > COOLING_THRESHOLD:
        return State.Cooling
    return State.Off

def print_status_to_console(sensor1: float, sensor2: float, sensor3: float, heater_state: State) -> None:
    print("Sensor1: " + str(sensor1) +
          ", Sensor2: " + str(sensor2) +
          ", Sensor3: " + str(sensor3) +
          ", Heater is " + heater_state.name)


def calc_temperature(value1: float, value2: float, value3: float) -> float:
    return (value1 + value2 + value3) / 3


def control_room_temperature():
    global write_to_console_counter

    sensor1 = get_sensor_value1()
    sensor2 = get_sensor_value2()
    sensor3 = get_sensor_value3()

    temperature = calc_temperature(sensor1, sensor2, sensor3)

    heater_state = get_heater_state(temperature)

    set_heater_state(heater_state)

    if write_to_console_counter >= 19:
        print_status_to_console(sensor1, sensor2, sensor3, heater_state)
        write_to_console_counter = 0
    else:
        write_to_console_counter += 1

File: reference/vector_samples_19_4_10/SIL/SilAdapterBasicCpp/SUT/Application_C/src/Logic.c
/*
 * This file contains the actual application logic.
 */

#include <stdio.h>

#include "IO.h"
#include "Logic.h"

#define HEATING_THRESHOLD 18.0
#define COOLING_THRESHOLD 23.0

enum State GetHeaterState(double temperature) {
  if (temperature < HEATING_THRESHOLD)
  {
    return STATE_HEATING;
  }

  if (temperature > COOLING_THRESHOLD)
  {
    return STATE_COOLING;
  }

  return STATE_OFF;
}

const char* GetHeaterStateString(enum State heaterState)
{
  switch (heaterState) {
  case STATE_OFF:
    return "Off";
  case STATE_HEATING:
    return "Heating";
  case STATE_COOLING:
    return "Cooling";
  default:
    return "Unknown State";
  }
}

void PrintStatusToConsole(double sensor1, double sensor2, double sensor3, enum State heaterState)
{
    printf("Sensor1: %.2f, Sensor2: %.2f, Sensor3: %.2f, Heater is %s\r\n",
      sensor1,
      sensor2,
      sensor3,
      GetHeaterStateString(heaterState));
}

double CalcTemperature(double value1, double value2, double value3) { return (value1 + value2 + value3) / 3.0; }

void ControlRoomTemperature()
{
  static int writeToConsoleCounter = 0;

  const double sensor1 = GetSensorValue1();
  const double sensor2 = GetSensorValue2();
  const double sensor3 = GetSensorValue3();

  const double temperature = CalcTemperature(sensor1, sensor2, sensor3);

  const enum State heaterState = GetHeaterState(temperature);

  SetHeaterState(heaterState);

  if (writeToConsoleCounter++ >= 19)
  {
    PrintStatusToConsole(sensor1, sensor2, sensor3, heaterState);
    writeToConsoleCounter = 0;
  }
}
